### PR TITLE
Add possibility to delete, cut and move selected tiles in tilemap

### LINKED ISSUE #9172: Add possibility to delete, cut and move selected tiles in tilemap
**Is your feature request related to a problem? Please describe (REQUIRED):**

Currently we only have a possibility to copy and place selected tiles somewhere on the tilemap, but no possibility to "cut" the tiles, so that we can "move" the selected region, so that if we would like to move the region, we would need to select empty tiles an "paint" them over the tiles we moved, something like:

tilemap_moving

**Describe the solution you'd like (REQUIRED):**
Easier approach would be to allow users to use some key (e.g. Delete) to remove selection and key combination (Ctrl+X) to cut the selection), so that selection is removed from tilemap, but the brush is still "having" the cutted region or a single tile, so you can paint it in other position.

**Describe alternatives you've considered (REQUIRED):**
Instead of the above described UX with additional keys being pressed while tile or region is selected in a tilemap, we can use additional changing key (currently `Shift` is used anyway to select tile(s) from tilemap), so maybe holding `Shift` + `Ctrl` would cause an action of "moving" the selected tile(s) in the tilemap.

**Additional context (OPTIONAL):**
https://www.reddit.com/r/defold/comments/1dqcl45/can_you_move_tiles_in_the_map_editor/

### PR DESCRIPTION
**Pull request by AGulev**
Added an ability to cut tiles using `Shitf+Ctrl` and erase tiles using `Shirt+Atl` in the tilemap editor.

  
CleanShot 2024-07-05 at 08 02 17

Fix https://github.com/defold/defold/issues/9172

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [x] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [x] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [x] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [x] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: editor/src/clj/editor/tile_map.clj [hunk 1]
```diff
@@ -113,6 +113,21 @@
...
+        x1 (inc (max sx ex))
+        y1 (inc (max sy ey))]
+    (loop [x x0
+           y y0
+           cell-map (transient cell-map)]
+      (if (< y y1)
+        (if (< x x1)
+          (recur (inc x) y (dissoc! cell-map (cell-index x y)))
+          (recur x0 (inc y) cell-map))
+        (persistent! cell-map)))))
```
Comment (AGulev) [inline editor/src/clj/editor/tile_map.clj:129]: new function to erase area

File: editor/src/clj/editor/tile_map.clj [hunk 2]
```diff
@@ -742,7 +757,9 @@
 
 (def color-shader (shader/make-shader ::color-shader pos-color-vert pos-color-frag))
 
-
+(def ^:private white-color (double-array (map #(/ % 255.0) [255 255 255])))
+(def ^:private red-color (double-array (map #(/ % 255.0) [255 0 0])))
+(def ^:private orange-color (double-array (map #(/ % 255.0) [255 140 0])))
```
Comment (AGulev) [inline editor/src/clj/editor/tile_map.clj:762]: all the colors moved out from code to be able to reuse

File: editor/src/clj/editor/tile_map.clj [hunk 3]
```diff
@@ -1124,7 +1144,7 @@
   [active-layer-renderable op op-select-start op-select-end current-tile tile-dimensions brush viewport texture-set-data gpu-texture]
   (when active-layer-renderable
     {pass/transparent (cond
-                        (= :select op)
+                        (contains? select-ops op)
```
Comment (AGulev) [inline editor/src/clj/editor/tile_map.clj:1147]: just to avoid `or` with 3 conditions

File: editor/src/clj/editor/tile_map.clj [hunk 4]
```diff
@@ -1135,18 +1155,22 @@
...
+     pass/outline (cond
+                    (contains? select-ops op) [{:world-transform (:world-transform active-layer-renderable)
+                                                :render-fn render-editor-select
+                                                :user-data {:start op-select-start
+                                                            :end op-select-end
+                                                            :tile-dimensions tile-dimensions
+                                                            :color (case op
+                                                                     :select-cut orange-color
+                                                                     :select-erase red-color
+                                                                     white-color)}}]
```
Comment (AGulev) [inline editor/src/clj/editor/tile_map.clj:1167]: olnly color added here

File: editor/src/clj/editor/tile_map.clj [hunk 5]
```diff
@@ -1191,42 +1215,85 @@
...
-      [(g/set-property self :brush (make-brush-from-selection cell-map start end))
-       (g/set-property self :op-select-start nil)
-       (g/set-property self :op-select-end nil)])))
+      (concat
+        (when brush-fn
+         [(g/set-property self :brush (brush-fn cell-map start end))])
+        (when erase?
+          [(g/update-property active-layer :cell-map erase-area start end)])
+        [(g/set-property self :op-select-start nil)
+         (g/set-property self :op-select-end nil)]))))
```
Comment (AGulev) [inline editor/src/clj/editor/tile_map.clj:1242]: the difference between functions, the rest are common

File: editor/src/clj/editor/tile_map.clj [hunk 6]
```diff
@@ -1191,42 +1215,85 @@
...
         tx (case (:type action)
              :mouse-pressed
              (when-not (some? op)
-               (let [op (if (true? (:shift action))
-                          :select
-                          :paint)
+               (let [op (cond
+                          (and (true? (:shift action)) (true? (:control action))) :select-cut
+                          (and (true? (:shift action)) (true? (:alt action))) :select-erase
+                          (true? (:shift action)) :select
```
Comment (AGulev) [inline editor/src/clj/editor/tile_map.clj:1295]: keys specified here

File: editor/src/clj/editor/tile_map.clj [hunk 7]
```diff
@@ -1191,42 +1215,85 @@
...
   (when-let [active-layer (g/node-value self :active-layer evaluation-context)]
     (let [cell-map (g/node-value active-layer :cell-map evaluation-context)
           start (g/node-value self :op-select-start evaluation-context)
           end (g/node-value self :op-select-end evaluation-context)]
-      [(g/set-property self :brush (make-brush-from-selection cell-map start end))
-       (g/set-property self :op-select-start nil)
-       (g/set-property self :op-select-end nil)])))
+      (concat
+        (when brush-fn
+         [(g/set-property self :brush (brush-fn cell-map start end))])
```
Comment (matgis) [inline editor/src/clj/editor/tile_map.clj:1238]: Fix indentation:
```suggestion
        (when brush-fn
          [(g/set-property self :brush (brush-fn cell-map start end))])
```

**Review by matgis — APPROVED** [review]
Works great, and a very nice addition! ❤️

As a future improvement, it would be nice if we could hide the brush and show the colored picking rectangle around the hovered tile immediately when the user holds down the picking modifier key.

Although I'm not sure the view event handler is notified when the user starts holding down a modifier key at the moment. It might require some tricky changes to the view to be able to do it. 🤔

Anyway, let's get this merged for now!

**Pull request merged**